tests/ranged_balance.cpp

```diff
@@ -0,0 +1,123 @@
+    shooter.wear_item( item( "backpack" ) );
+
+    // Give shooter a loaded gun of the requested type.
+    item &gun = shooter.i_add( item( gun_id ) );
+    const itype_id ammo_id = gun.ammo_default();
+    if( gun.magazine_integral() ) {
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, gun.ammo_capacity() ) );
+	REQUIRE( gun.is_reloadable_with( ammo_id ) );
+	REQUIRE( shooter.can_reload( gun, ammo_id ) );
+	gun.reload( shooter, item_location( shooter, &ammo ), gun.ammo_capacity() );
+    } else {
+        const itype_id magazine_id = gun.magazine_default();
+        item &magazine = shooter.i_add( item( magazine_id ) );
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, magazine.ammo_capacity() ) );
+        REQUIRE( magazine.is_reloadable_with( ammo_id ) );
+        REQUIRE( shooter.can_reload( magazine, ammo_id ) );
+        magazine.reload( shooter, item_location( shooter, &ammo ), magazine.ammo_capacity() );
+        gun.reload( shooter, item_location( shooter, &magazine ), magazine.ammo_capacity() );
+    }
+    shooter.wield( gun );
+}
+
+static statistics firing_test( npc &shooter, std::string gun_type, float aim_ratio, int range )
+{
+    equip_shooter( shooter, itype_id( gun_type ) );
+
+    // Spawn a target.
+    // Maybe change this to a random direction, or even move the target around during the test.
+    tripoint target_pos( shooter.pos() + tripoint( range, 0, 0 ) );
+    monster &test_target = spawn_test_monster( "debug_mon", target_pos );
+
+    item &gun = shooter.weapon;
+
+    dispersion_sources dispersion =
+      shooter.get_weapon_dispersion( gun,
+                                     rl_dist( shooter.pos(), test_target.pos() ) );
+    dispersion.add_range( std::max( static_cast<double>( gun.sight_dispersion() ),
+                                    shooter.recoil_total() * ( 1.0 - aim_ratio ) ) );
+
+    statistics firing_stats;
+    do {
+        // Top off target hp to make sure it doesn't die.
+        test_target.set_hp( test_target.get_hp_max() );
+        // Calling projectile_ttack() instead of fire_gun() because it returns wether it was a hit or not.
+        // Conviniently, it also avoids consuming the ammo.
+        dealt_projectile_attack shot =
+          projectile_attack( make_gun_projectile( gun ), shooter.pos(), test_target.pos(),
+                             dispersion, &shooter, nullptr );
+        firing_stats.add( shot.hit_critter != nullptr );
+    } while( firing_stats.n() < 100 ||
+             ( firing_stats.n() < 1000 && firing_stats.adj_wald_error() < 0.0001 ) );
+    return firing_stats;
+}
+
+static void test_shooting_scenario( npc &shooter, std::string gun_type,
+                                    int minimum_range, int good_range )
+{
+    clear_map();
+
+    statistics minimum_stats = firing_test( shooter, gun_type, 0.0, minimum_range );
+    INFO( "Accumulated " << minimum_stats.n() << " samples." );
+    CHECK( minimum_stats.avg() < 0.01 );
```

Comment:
Why not? If it starts failing we can adjust, but in this case we're asserting "almost never hits", so I'd like to be as strong as is practical.
Note I'm not rejecting your request to factor in hit quality, I'm just filling this out in what I think is the best order, I'll circe back around to check if hit quality is practical to extract once I have the use cases covered.